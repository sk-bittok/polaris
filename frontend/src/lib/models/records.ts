export interface ProductionRecord {
	id: number;
	livestockName: string;
	animalPid: string;
	productType: string;
	unit: string;
	quantity: string;
	notes?: string;
	quality?: string;
	recordDate?: Date;
	createdAt: Date;
	createdBy: string;
	createdByName: string;
	updatedAt: Date;
}

export interface NewProductionRecord {
	id: number;
	animal_pid: string;
	organisation_pid: string;
	product_type: string;
	quantity: number;
	unit: string;
	record_date?: string;
	quality?: string;
	notes?: string;
	created_by: string;
	created_at: Date;
	updated_at: Date;
}

export interface HealthRecordResponse {
	id: number;
	organisationPid: string;
	organisationName: string;
	animalName: string;
	animalPid: string;
	animalTagId: string;
	recordType: string;
	recordDate: Date;
	description: string;
	treatment: string;
	medicine?: string;
	dosage?: string;
	cost?: number;
	notes?: string;
	performedBy?: string;
	createdBy: string;
	createdByName: string;
	createdAt: Date;
	updatedAt: Date;
}

export interface HealthRecord {
	id: number;
	organisation_pid: string;
	animal_pid: string;
	record_type: string;
	record_date: Date;
	description: string;
	treatment: string;
	medicine?: string;
	dosage?: string;
	cost?: number;
	notes?: string;
	performed_by?: string;
	created_by: string;
	created_at: Date;
	updated_at: Date;
}

export interface WeightRecord {
	id: number;
	organisation_pid: string;
	animal_pid: string;
	mass: number;
	record_date: string;
	notes?: string;
	created_by: string;
	created_at: Date;
	updated_at: Date;
}

export interface WeightRecordResponse {
	id: number;
	organisationPid: string;
	organisationName: string;
	animalPid: string;
	animalName: string;
	animalTagId: string;
	mass: number;
	recordDate: string;
	notes?: string;
	createdBy: string;
	createdByName: string;
	created_at: Date;
	updated_at: Date;
}
