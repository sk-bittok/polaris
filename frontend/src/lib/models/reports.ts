interface LivestockSummary {
	pid: string;
	organisationPid: string;
	id: number;
	total: number;
	males: number;
	females: number;
	unkownGender: number;
	active: number;
	transferred: number;
	sold: number;
	deceased: number;
	species: number;
	breeds: number;
	totalPurchasedValue: number;
	createdAt: Date;
	updatedAt: Date;
}

export type { LivestockSummary };
