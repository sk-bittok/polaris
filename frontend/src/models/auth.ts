import { z } from "zod";

export enum SubscriptionType {
    Basic = "basic",
    Business = "business",
    Enterprise = "enterprise",
}

export const companyFormSchema = z.object({
    companyName: z.string().min(2, { message: "Name requires 2 characters" })
        .max(200, { message: "Name is too long" }),
    companyAddress: z.string().max(200, { message: "Address is too long" })
        .nullable(),
    companyPhone: z.string().min(10, { message: "Phone is too short" })
        .max(12, { message: "Phone is too long" }).nullable(),
    companyEmail: z.string().email({ message: "Invalid email" }).nullable(),
    subscriptionType: z.enum(["basic", "business", "enterprise"], {
        message: "Basic, business and enterprise only",
    }),
});

export const userFormSchema = z.object({
    email: z.string().email({ message: "Invalid email" }),
    firstName: z.string().min(2, { message: "Name is too short" })
        .max(50, { message: "Name is too long" }),
    lastName: z.string().min(2, { message: "Name is too short" }).max(50, {
        message: "Name is too long",
    }),
    password: z.string().min(8, { message: "Password requires 8 characters" })
        .max(48, { message: "Password must be under 48 chars" }),
    confirmPassword: z.string(),
})
    .refine((data) => data.confirmPassword === data.password, {
        message: "Passwords must match",
        path: ["confirmPassword"],
    });

export const loginFormSchema = z.object({
    email: z.string().email({ message: "Invalid email" }),
    password: z.string().min(2, { message: "Password too short" })
        .max(100, { message: "Password too long" }),
});

export type CompanyForm = z.infer<typeof companyFormSchema>;
export type UserForm = z.infer<typeof userFormSchema>;

export type RegisterOrgAndUser = {
    organisation: CompanyForm;
    user: UserForm;
};

export type LoginForm = z.infer<typeof loginFormSchema>;
