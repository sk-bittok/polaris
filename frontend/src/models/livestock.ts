
export interface Breed {
  id: number;
  organisationPid?: string;
  specie: Category;
  name: string;
  maleWeightRange?: string;
  femaleWeightRange?: string;
  gestationPeriod?: string;
  description?: string;
  createdAt: string;
  isSystemDefined: boolean;
}

export enum Category {
  Cattle = "cattle",
  Sheep = "sheep",
  Goat = "goats",
  Chicken = "chicken",
  Pig = "pigs"
}

export enum Gender {
  Male = "male",
  Female = "female",
  Unkown = "unkown",
}

export enum Status {
  Active = "active",
  Sold = "sold",
  Deceased = "deceased",
  Transferred = "transferred"
}

export interface RegisterLivestock {
  tagId: string;
  name: string;
  gender: string;
  status: string;
  breed: string;
  specie: string;
  dateOfBirth?: string | Date | null;
  weightAtBirth?: number | null;
  maleParentId?: string | null;
  femaleParentId?: string | null;
  purchasePrice?: number | null;
  purchaseDate?: string | Date | null;
  currentWeight?: number | null;
  notes?: string | null;
};

export interface UpdateLivestock {
  tagId?: string | null;
  name?: string | null;
  gender?: string | null;
  status?: string | null;
  breed?: string | null;
  specie?: string | null;
  dateOfBirth?: string | Date | null;
  weightAtBirth?: number | null;
  maleParentId?: string | null;
  femaleParentId?: string | null;
  purchasePrice?: number | null;
  purchaseDate?: string | Date | null;
  currentWeight?: number | null;
  notes?: string | null;
};

export interface Livestock {
  id: number;
  pid: string;
  organisationName: string;
  tagId: string;
  name: string;
  breedName: string;
  specieName: Category;
  dateOfBirth?: Date;
  gender: Gender;
  parentMaleName?: string;
  parentFemaleName?: string;
  status: Status;
  parentMaleTagId?: string;
  parentFemaleTagId?: string;
  purchaseDate?:  Date;
  purchasePrice?: string;
  weightAtBirth?: string;
  currentWeight?: string;
  notes?: string;
  createdByName: string;
  createdAt: Date;
  updatedAt: Date;
}

