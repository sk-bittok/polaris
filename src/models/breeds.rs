#![allow(clippy::missing_errors_doc)]

use std::borrow::Cow;

use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, prelude::FromRow};
use uuid::Uuid;

use super::{ModelError, ModelResult};

#[derive(Debug, Clone, Deserialize)]
pub struct BreedQuery {
    specie: Option<String>,
    name: Option<String>,
}

#[derive(Debug, Deserialize, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegisterBreed<'a> {
    pub(crate) specie: Cow<'a, str>,
    pub(crate) name: Cow<'a, str>,
    pub(crate) description: Option<Cow<'a, str>>,
    pub(crate) typical_male_weight_range: Option<Cow<'a, str>>,
    pub(crate) typical_female_weight_range: Option<Cow<'a, str>>,
    pub(crate) typical_gestation_period: Option<Cow<'a, str>>,
}

#[derive(Debug, Deserialize, Serialize, Clone, FromRow, Encode)]
pub struct Breed {
    pub(crate) id: i32,
    pub(crate) specie: String,
    pub(crate) name: String,
    pub(crate) description: Option<String>,
    pub(crate) typical_male_weight_range: Option<String>,
    pub(crate) typical_female_weight_range: Option<String>,
    pub(crate) typical_gestation_period: Option<String>,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) organisation_pid: Option<Uuid>,
    pub(crate) is_system_defined: bool,
}

impl Breed {
    pub async fn find_by_id<'e, C>(db: C, id: i32, org_pid: Uuid) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE id = $1 AND (is_system_defined = TRUE OR organisation_pid = $2)")
            .bind(id)
            .bind(org_pid)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_all<'e, C>(
        db: C,
        org_pid: Uuid,
        conditions: &BreedQuery,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let mut query =
            sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE  (is_system_defined = TRUE OR organisation_pid = $1) ORDER BY name")
                .bind(org_pid);

        if let Some(name) = &conditions.name {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE (is_system_defined = TRUE OR organisation_pid = $1) AND name LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{name}%"));
        }

        if let Some(specie) = &conditions.specie {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE organisation_pid = $1 AND specie LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{specie}%",));
        }

        query.fetch_all(db).await.map_err(Into::into)
    }

    pub async fn create<'e, C>(db: C, org_pid: Uuid, dto: &RegisterBreed<'_>) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let query = sqlx::query_as::<_, Self>(
            "INSERT INTO breeds (specie, name, typical_male_weight_range, typical_female_weight_range, typical_gestation_period, organisation_pid, description)
                VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *
            ")
        .bind(&dto.specie)
        .bind(&dto.name)
        .bind(&dto.typical_male_weight_range)
        .bind(&dto.typical_female_weight_range)
        .bind(&dto.typical_gestation_period)
        .bind(org_pid)
        .bind(&dto.description)
        .fetch_one(db)
        .await?;

        Ok(query)
    }
}
