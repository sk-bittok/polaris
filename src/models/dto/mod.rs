pub mod auth;

pub use self::auth::*;
