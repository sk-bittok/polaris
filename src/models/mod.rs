pub mod animals;
pub mod breeds;
pub mod dto;
pub mod enums;
pub mod errors;
pub mod health;
pub mod livestock;
pub mod orgs;
pub mod production;
pub mod roles;
pub mod species;
pub mod summaries;
pub mod users;
pub mod weight;

pub use self::{
    errors::{ModelError, ModelResult},
    summaries::*,
};
