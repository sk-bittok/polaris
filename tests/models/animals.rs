#![allow(unused)]
use std::{borrow::Cow, str::FromStr};

use chrono::NaiveDate;
use insta::{Settings, assert_debug_snapshot, with_settings};
use polaris::models::{
    animals::{Animal, AnimalQuery},
    dto::{Gender, LinkOffspring, RegisterAnimal, UpdateAnimal},
};
use rstest::rstest;
use serial_test::serial;
use uuid::Uuid;

use crate::{boot_test, cleanup_date, cleanup_int, cleanup_uuid, seed_data};

macro_rules! configure_insta {
    ($(expr:expr),*) => {
        let mut settings = Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("animals");
        settings.set_snapshot_path("snapshots/animals");
        let _guard = settings.bind_to_scope();
    };
}

#[rstest]
#[case(
    "can_find_all_no_query",
    Uuid::parse_str( "9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap(),
     AnimalQuery {
        specie: None,
        breed: None,
        purchase_date: None,
        female_parent: None,
        male_parent: None,
    }
)]
#[case(
    "can_find_all_specie_query",
    Uuid::parse_str( "9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap(),
     AnimalQuery {
        specie: Some("cattle".to_string()),
        breed: None,
        purchase_date: None,
        female_parent: None,
        male_parent: None,
    }
)]
#[case(
    "can_find_all_breed_query",
    Uuid::parse_str("4a0f3af9-e56e-4e21-8f3a-f9e56efe215b").unwrap(),
     AnimalQuery {
        specie: None,
        breed: Some("Merino".into()),
        purchase_date: None,
        female_parent: None,
        male_parent: None,
    }
)]
#[case(
    "can_find_all_female_parent_query",
    Uuid::parse_str("4a0f3af9-e56e-4e21-8f3a-f9e56efe215b").unwrap(),
     AnimalQuery {
        specie: None,
        breed: None,
        purchase_date: None,
        female_parent: Some(Uuid::parse_str("487a7b25-3ea9-4a40-ae8c-43cdf51138be").unwrap()),
        male_parent: None,
    }
)]
#[case(
    "can_find_all_male_parent_query",
    Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap(),
     AnimalQuery {
        specie: None,
        breed: None,
        purchase_date: None,
        male_parent: Some(Uuid::parse_str("d909e761-36da-4062-ae78-abba4f7c1103").unwrap()),
        female_parent: None,
    }
)]
#[tokio::test]
#[serial]
async fn can_find_all(#[case] name: &str, #[case] org_pid: Uuid, #[case] conditions: AnimalQuery) {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let result = Animal::find_all(&ctx.db, org_pid, &conditions).await;

    assert_debug_snapshot!(name, result);
}

#[rstest]
#[case(
    "can_find_by_id",
    Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap(),
     Uuid::parse_str("b2bd6270-8bec-42ce-99ff-d0eb1a076221").unwrap(),
)]
#[case(
    "can_find_by_id_wrong_id",
    Uuid::parse_str("4a93f0a8-4a91-482d-92d8-f0b3b084c2e4").unwrap(),
     Uuid::parse_str("b2bd6270-8bec-42c4-99ff-d0eb1a076221").unwrap(),
)]
#[tokio::test]
#[serial]
async fn can_find_by_id(#[case] test_name: &str, #[case] org_pid: Uuid, #[case] id: Uuid) {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let result = Animal::find_by_id(&ctx.db, org_pid, id).await;

    assert_debug_snapshot!(test_name, result);
}

#[tokio::test]
#[serial]
async fn can_create_one() {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let user_pid = Uuid::parse_str("e761d8e3-fc3e-4a2e-a6c9-7c7a4f2130e8").unwrap();
    let org_pid = Uuid::parse_str("4a93f0a8-4a91-482d-92d8-f0b3b084c2e4").unwrap();

    let params = RegisterAnimal {
        tag_id: Cow::Borrowed("GX027"),
        name: Cow::Borrowed("Monkey Blue"),
        gender: Cow::Borrowed("male"),
        status: Cow::Borrowed("active"),
        specie: Cow::Borrowed("cattle"),
        breed: Cow::Borrowed("beef shorthorn"),
        date_of_birth: Some(Cow::Borrowed("2023-12-17")),
        female_parent_id: None,
        male_parent_id: None,
        purchase_date: None,
        purchase_price: None,
        weight_at_birth: Some(3600),
        current_weight: None,
        notes: None,
    };

    let result = Animal::register(&ctx.db, org_pid, user_pid, &params).await;

    with_settings!({
        filters => {
            let mut filters = cleanup_date().to_vec();
            filters.extend(cleanup_uuid().to_vec());
            filters.extend(cleanup_int().to_vec());
            filters
        }
    }, {
        assert_debug_snapshot!(result);
        });
}

#[tokio::test]
#[serial]
async fn can_create_one_with_parents() {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let user_pid = Uuid::parse_str("bd6f7c26-d2c9-487e-b837-8f77be468033").unwrap();
    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let params = RegisterAnimal {
        tag_id: Cow::Borrowed("GX027"),
        name: Cow::Borrowed("Josephone"),
        gender: Cow::Borrowed("female"),
        status: Cow::Borrowed("active"),
        specie: Cow::Borrowed("cattle"),
        breed: Cow::Borrowed("beef shorthorn"),
        date_of_birth: Some(Cow::Borrowed("2023-12-25")),
        female_parent_id: Some(Cow::Borrowed("AC001")),
        male_parent_id: Some(Cow::Borrowed("AC003")),
        purchase_date: None,
        purchase_price: None,
        weight_at_birth: Some(3600),
        current_weight: Some(55000),
        notes: None,
    };

    let result = Animal::register(&ctx.db, org_pid, user_pid, &params).await;

    with_settings!({
        filters => {
            let mut filters = cleanup_date().to_vec();
            filters.extend(cleanup_uuid().to_vec());
            filters.extend(cleanup_int().to_vec());
            filters
        }
    }, {
        assert_debug_snapshot!(result);
        });
}

#[tokio::test]
#[serial]
async fn can_update_one() {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let user_pid = Uuid::parse_str("bd6f7c26-d2c9-487e-b837-8f77be468033").unwrap();
    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();
    let id = Uuid::parse_str("b2bd6270-8bec-42ce-99ff-d0eb1a076221").unwrap();

    let params = UpdateAnimal {
        name: Some(Cow::Borrowed("Gertrude")),
        tag_id: None,
        breed: Some(Cow::Borrowed("Friesian")),
        gender: None,
        status: None,
        specie: None,
        date_of_birth: Some(Cow::Borrowed("2022-7-9")),
        female_parent_id: None,
        male_parent_id: None,
        purchase_date: None,
        purchase_price: None,
        weight_at_birth: None,
        current_weight: Some(61000),
        notes: None,
    };

    let result = Animal::update_by_id(&ctx.db, &params, org_pid, id).await;

    with_settings!({
        filters => {
            crate::cleanup_date().to_vec()
        }
    }, {
        assert_debug_snapshot!(result);
    })
}

#[tokio::test]
#[serial]
async fn can_find_by_tag_id() {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();
    let tag_id = "AC003";

    let result = Animal::find_by_tag_id(&ctx.db, org_pid, tag_id).await;

    assert_debug_snapshot!(result);
}

#[rstest]
#[case(
    "can_link_offspring_male_parent",
    LinkOffspring {
        offspring_tag_id: Cow::Borrowed("AC010"),
        offspring_name: Cow::Borrowed("Duke"),
        parent_tag_id: Cow::Borrowed("AC003"),
        parent_gender: Gender::Male
    }
)]
#[case(
    "can_link_offspring_female_parent",
    LinkOffspring {
        offspring_tag_id: Cow::Borrowed("AC010"),
        offspring_name: Cow::Borrowed("Duke"),
        parent_tag_id: Cow::Borrowed("AC001"),
        parent_gender: Gender::Female
    }
)]
#[tokio::test]
#[serial]
async fn can_link_offspring(#[case] test_name: &str, #[case] params: LinkOffspring<'_>) {
    configure_insta!();

    let ctx = boot_test().await.unwrap();
    seed_data(&ctx.db).await.unwrap();

    let org_pid = Uuid::parse_str("9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0").unwrap();

    let result = Animal::link_offspring(&ctx.db, org_pid, &params).await;

    with_settings!({
        filters => {
            crate::cleanup_date().to_vec()
        }
    }, {
        assert_debug_snapshot!(test_name, result);
    })
}
