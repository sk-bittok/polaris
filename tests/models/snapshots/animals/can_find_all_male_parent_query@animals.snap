---
source: tests/models/animals.rs
assertion_line: 92
expression: result
---
Ok(
    [
        AnimalResponse {
            id: 127,
            pid: 037c7aeb-b8e5-4900-bdd8-cf24a848a3a5,
            organisation_pid: 9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0,
            organisation_name: "Acme Corp",
            tag_id: "AC011",
            name: "Clover",
            specie_name: "cattle",
            breed_name: "Jersey",
            date_of_birth: Some(
                2024-04-10,
            ),
            gender: "female",
            parent_female_name: Some(
                "Buttercup",
            ),
            parent_female_tag_id: Some(
                "AC002",
            ),
            parent_female_id: Some(
                62197c29-a2dd-4591-a4d0-92f6d7099760,
            ),
            parent_male_name: Some(
                "Ferdinand",
            ),
            parent_male_tag_id: Some(
                "AC003",
            ),
            parent_male_id: Some(
                d909e761-36da-4062-ae78-abba4f7c1103,
            ),
            status: "active",
            purchase_date: None,
            purchase_price: None,
            weight_at_birth: Some(
                26.75,
            ),
            current_weight: Some(
                145.50,
            ),
            notes: Some(
                "Twin of Rose, slightly smaller but healthy",
            ),
            created_by: bd6f7c26-d2c9-487e-b837-8f77be468033,
            created_by_name: "John Doe",
            created_at: 2024-04-10T07:25:00+00:00,
            updated_at: 2024-12-08T16:45:00+00:00,
        },
        AnimalResponse {
            id: 115,
            pid: f6417c11-d817-4626-9e8d-c68a44002d4b,
            organisation_pid: 9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0,
            organisation_name: "Acme Corp",
            tag_id: "AC007",
            name: "Rose",
            specie_name: "cattle",
            breed_name: "Jersey",
            date_of_birth: Some(
                2024-04-10,
            ),
            gender: "female",
            parent_female_name: Some(
                "Spot",
            ),
            parent_female_tag_id: Some(
                "AC005",
            ),
            parent_female_id: Some(
                5a6efa8e-8cf3-46fb-9fe6-41900aca729b,
            ),
            parent_male_name: Some(
                "Ferdinand",
            ),
            parent_male_tag_id: Some(
                "AC003",
            ),
            parent_male_id: Some(
                d909e761-36da-4062-ae78-abba4f7c1103,
            ),
            status: "active",
            purchase_date: None,
            purchase_price: None,
            weight_at_birth: Some(
                27.25,
            ),
            current_weight: Some(
                150.75,
            ),
            notes: Some(
                "Healthy calf, growing well",
            ),
            created_by: bd6f7c26-d2c9-487e-b837-8f77be468033,
            created_by_name: "John Doe",
            created_at: 2024-04-10T07:15:00+00:00,
            updated_at: 2024-12-10T11:30:00+00:00,
        },
        AnimalResponse {
            id: 104,
            pid: bf3a3228-2188-4af4-ad55-0184cd5c3449,
            organisation_pid: 9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0,
            organisation_name: "Acme Corp",
            tag_id: "AC004",
            name: "Bella",
            specie_name: "cattle",
            breed_name: "Jersey",
            date_of_birth: Some(
                2024-02-18,
            ),
            gender: "female",
            parent_female_name: Some(
                "Daisy",
            ),
            parent_female_tag_id: Some(
                "AC001",
            ),
            parent_female_id: Some(
                b2bd6270-8bec-42ce-99ff-d0eb1a076221,
            ),
            parent_male_name: Some(
                "Ferdinand",
            ),
            parent_male_tag_id: Some(
                "AC003",
            ),
            parent_male_id: Some(
                d909e761-36da-4062-ae78-abba4f7c1103,
            ),
            status: "active",
            purchase_date: None,
            purchase_price: None,
            weight_at_birth: Some(
                26.50,
            ),
            current_weight: Some(
                180.25,
            ),
            notes: Some(
                "First calf born on the farm",
            ),
            created_by: bd6f7c26-d2c9-487e-b837-8f77be468033,
            created_by_name: "John Doe",
            created_at: 2024-02-18T07:00:00+00:00,
            updated_at: 2024-12-05T13:15:00+00:00,
        },
    ],
)
